File: src/main/java/net/dontdrinkandroot/gitki/wicket/dropdown/OptionLinkDropdown.java
package net.dontdrinkandroot.gitki.wicket.dropdown;

import net.dontdrinkandroot.gitki.wicket.css.GitkiCssClass;
import net.dontdrinkandroot.wicket.behavior.CssClassAppender;
import net.dontdrinkandroot.wicket.bootstrap.behavior.DropdownToggleBehavior;
import net.dontdrinkandroot.wicket.bootstrap.behavior.IconBehavior;
import net.dontdrinkandroot.wicket.bootstrap.component.dropdown.DropdownMenu;
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass;
import net.dontdrinkandroot.wicket.bootstrap.css.DropdownAlignment;
import net.dontdrinkandroot.wicket.css.StringCssClass;
import org.apache.wicket.Component;
import org.apache.wicket.markup.html.WebMarkupContainer;
import org.apache.wicket.markup.html.panel.Panel;
import org.apache.wicket.markup.repeater.RepeatingView;

public abstract class OptionLinkDropdown extends Panel
{
    public OptionLinkDropdown(String id)
    {
        super(id);
    }

    @Override
    protected void onInitialize()
    {
        super.onInitialize();
        this.add(new CssClassAppender(BootstrapCssClass.DROPDOWN));

        Component link = new WebMarkupContainer("link");
        link.add(new IconBehavior(new StringCssClass("fas fa-ellipsis-v fa-fw")));
        link.add(new DropdownToggleBehavior());
        link.add(new CssClassAppender(GitkiCssClass.CARET_OFF));
        this.add(link);

        DropdownMenu dropdownMenu = new DropdownMenu("dropdownMenu")
        {
            @Override
            protected void populateItems(RepeatingView itemView)
            {
                OptionLinkDropdown.this.populateItems(itemView);
            }
        };
        dropdownMenu.setAlignment(DropdownAlignment.RIGHT);
        this.add(dropdownMenu);
    }

    protected abstract void populateItems(RepeatingView itemView);
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/dropdown/OptionLinkDropdown.kt
package net.dontdrinkandroot.gitki.wicket.dropdown

import net.dontdrinkandroot.wicket.behavior.CssClassAppender
import net.dontdrinkandroot.wicket.bootstrap.behavior.DropdownToggleBehavior
import net.dontdrinkandroot.wicket.bootstrap.behavior.IconBehavior
import net.dontdrinkandroot.wicket.bootstrap.component.dropdown.DropdownMenu
import net.dontdrinkandroot.wicket.bootstrap.component.item.ItemView
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass
import net.dontdrinkandroot.wicket.bootstrap.css.DropdownAlignment
import net.dontdrinkandroot.wicket.bootstrap.css.FontAwesome5IconClass
import net.dontdrinkandroot.wicket.markup.html.markupContainer
import org.apache.wicket.markup.html.panel.Panel
import org.apache.wicket.markup.repeater.RepeatingView
import org.apache.wicket.model.Model

class OptionLinkDropdown(
    id: String,
    populateItemsHandler: OptionLinkDropdown.(itemView: RepeatingView) -> Any?
) : Panel(id) {

    init {
        this.add(CssClassAppender(BootstrapCssClass.DROPDOWN))
        markupContainer(
            "link",
            DropdownToggleBehavior(),
            IconBehavior(FontAwesome5IconClass.ELLIPSIS_V.createIcon(fixedWidth = true))
        )
        this.add(object : DropdownMenu("dropdownMenu", Model(DropdownAlignment.END)) {
            override fun populateItems(itemView: ItemView) {
                populateItemsHandler(itemView)
            }
        })
    }
}